src/Result/Warning.php

```diff
@@ -22,12 +20,12 @@ private function __construct(string $message)
      * @param string[] $layerNames
      */
     public static function tokenIsInMoreThanOneLayer(
-        TokenInterface $tokenName,
```

Comment:
This is similar to my complaint in the OutputFormatter. We leak an implementation detail for rendering the output to the callee. In this case it's even more clear, why this is bad. When we switch to PHP 8+ later and extend Stringable for the TokenInterface, then we can make a string out of each token without having to call `->toString()`. The current change will create a lot of noise in the PR, whereas the old implementation would only affect this function.

Reply:
The problem here is now we are leaking implementation details of the engine (`Ast`'s `TokenInterface`) to the public `Contract` API. Here you should know nothing about how we internally get the information about the token name.